File: 프로그래머스/unrated/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        int dpointer=n-1;
        int ppointer=n-1;
        long answer=0;
        while(dpointer>=0||ppointer>=0){
            while(dpointer>=0&&deliveries[dpointer]==0) --dpointer;
            while(ppointer>=0&&pickups[ppointer]==0) --ppointer;
            answer+=(Math.max(dpointer,ppointer)+1)*2;
            int sum=0;
            while(dpointer>=0&&cap>sum){
                sum+=deliveries[dpointer];
                deliveries[dpointer--]=0;
            }
            if(cap<sum){
                deliveries[++dpointer]=sum-cap;
            }
            sum=0;
            while(ppointer>=0&&cap>sum){
                sum+=pickups[ppointer];
                pickups[ppointer--]=0;
            }
            if(cap<sum){
                pickups[++ppointer]=sum-cap;
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.cpp
#include <bits/stdc++.h>

using namespace std;
vector<int> d;
vector<int> p;
long long solution(int cap, int n, vector<int> deliveries, vector<int> pickups) {
    long long answer = 0;
    vector<int> d, p;

    for(int i=0; i<n; i++){
        if(deliveries[i] > 0) d.push_back(i);
        if(pickups[i] > 0) p.push_back(i);
    }
    
    while(!d.empty() || !p.empty()){
        int d_distance = d.empty() ? 0 : d.back()+1;
        int p_distance = p.empty() ? 0 : p.back()+1;
        int distance = max(d_distance, p_distance);

        int amount = cap;

        // 배달
        while(amount && !d.empty()){
            int idx = d.back();
            if(deliveries[idx] <= amount){
                amount -= deliveries[idx];
                deliveries[idx] = 0;
                d.pop_back();
            } else {
                deliveries[idx] -= amount;
                amount = 0;
            }
        }

        amount = cap;

        // 수거
        while(amount && !p.empty()){
            int idx = p.back();
            if(pickups[idx] <= amount){
                amount -= pickups[idx];
                pickups[idx] = 0;
                p.pop_back();
            } else {
                pickups[idx] -= amount;
                amount = 0;
            }
        }

        answer += (long long)distance * 2;
    }
    return answer;
}